### ISSUE TITLE
Return the ability to import realms using a env variable in the Docker

### ISSUE THREAD

**Issue by eabykov** [keycloak/keycloak#10216]
### Description

Need to return the ability to import realm at the start of the application to simplify its initial configuration

### Discussion

[Export/Import of realm data (JSON) #10229](https://github.com/keycloak/keycloak/discussions/10229)

### Motivation

Now we have to wait until Keycloak starts and do the import using one of the ways

1. Using `kcadm.sh`
```shell
/opt/keycloak/bin/kcadm.sh config credentials --server http://localhost:8080/auth --realm master --user ${KEYCLOAK_ADMIN} --password ${KEYCLOAK_ADMIN_PASSWORD}
/opt/keycloak/bin/kcadm.sh create realms -f /tmp/realm.json'
```
2. Using new `kc.sh import` which requires restarting Keycloak
```shell
/opt/keycloak/bin/kc.sh import --file /tmp/realm.json
```

Please simplify our life and return the opportunity to make import realm from a file using a startup variable because I try to create a guide on how to simple [run Keycloak in Docker Compose v2](https://github.com/eabykov/keycloak-compose) and this problem prevents the automation of the process

### Details

In the Keycloak 16.1.1 and previous versions
```yaml
  keycloak:
    image: "quay.io/keycloak/keycloak:16.1.1"
    environment:
      - KEYCLOAK_IMPORT=/tmp/realm.json
    volumes:
      - ./realm.json:/tmp/realm.json:ro
```

In the Keycloak 17.0.0, it doesn't work even though I tried different variables and there are no errors in the log either

```yaml
  keycloak:
    image: "quay.io/keycloak/keycloak:17.0.0"
    environment:
      - KEYCLOAK_IMPORT=/tmp/realm.json
      - KC_IMPORT=/tmp/realm.json
    volumes:
      - ./realm.json:/tmp/realm.json:ro
```

Duplicate of #9261

**Comment by vkalachevskyi** [comment]
You can create credentials with new env variables KEYCLOAK_ADMIN and KEYCLOAK_ADMIN_PASSWORD. But I have the same question about the realm.

**Comment by pedroigor** [comment]
@eabykov That capability is missing. The reason is that we were not sure if we should keep it as it is not something you do in production.

Do you agree this capability is more related to dev/testing? Would make sense to allow importing realms at startup only when in development mode?

**Comment by eabykov** [comment]
> @eabykov That capability is missing. The reason is that we were not sure if we should keep it as it is not something you do in production.

@pedroigor we need to automatically prepare the environment for clients to work, which means that

- clients
- users
- roles
- Tokens timings

Must be created at the deployment step. It would be very convenient to perform the initial setup without a bunch of API requests, but simply by inserting json and setting the environment variable.

PS: When we didn't know about the possibility of importing a realm via a file and an environment variable, we still imported it only through the API

> Do you agree this capability is more related to dev/testing? Would make sense to allow importing realms at startup only when in development mode?

I do not agree with you at this point, do you suggest that customers in the Prod solution create clients themselves or configure Tokens timings? I think it is necessary to automate the creation of realm not only in the dev, but also in prod

We have a set of specific applications that require pre-configured clients and roles so that the client only creates a new user and defines a role for him.
PS: We also have a set of users who exist by default, they also need to be created (for example, with the requirement to confirm email or by setting a temporary password etc.)

All this is necessary so that we can raise the prod environment for users in which they will execute a command such as `docker compose up -in --wait` and as a result, when clicking on the url to their application, they will be able to log in with a specific user by changing the temporary password and configuring OTP, and should not execute scripts with API requests

**Comment by dasniko** [comment]
@pedroigor 

I didn't see this issue and opened discussion #10229.

> Do you agree this capability is more related to dev/testing? Would make sense to allow importing realms at startup only when in development mode?

IMHO, you should provide _at least_ an auto-import feature on startup in dev mode! This would solve many challenges in development and testing scenarios.
Currently, in the new version of [testcontainers-keycloak](https://github.com/dasniko/testcontainers-keycloak) project, I'm doing the realm import via the Keycloak admin client, before, I just used the auto-import feature. In this example, it's ok to do it this way, because I have control over the lifecycle. In other scenarios, devs might not have that much control and/or might not be that familiar with controling lifecycle. Thus, such a feature definitely makes sense!

For production, I've still no clear opinion. But the longer/more I think about it, the less I need it in my supported environments (several customers). I used auto import in the early days also in production for new environments, but I never used it recently. Most recent scenarios, I used to write some update/config scripts with the Admin REST API via the Java admin client or shell scripts with `kcadm.sh`.

IMO, for production environments, it would be much much better to have something like [adorsys/keycloak-config-cli](https://github.com/adorsys/keycloak-config-cli/) supported natively / ootb by Keycloak itself.

**Comment by pedroigor** [comment]
> 
> @pedroigor we need to automatically prepare the environment for clients to work, which means that

I totally understand your points. And I see how the capability fits here. I'm open to supporting both approaches (api and startup) but my main concern is to keep consistency and rely on our API only. Also because the API offers you much more possibilities in terms of provisioning. For instance, check whether the realm is duplicated or if the state changed, etc.

Is your use case related to Keycloak as a SaaS? And how to provide the basis for starting using customer-owned Keycloak instances?

**Comment by pedroigor** [comment]
@MrBuddyCasino But if you are deploying to staging you are probably using data closer to production. And I would expect a lot of data to be imported and becoming sometimes too cumbersome to do.

That is my main argument for not having the capability in production/staging/or even testing.

It does make sense for provisioning though as per some use cases herein provided.

**Comment by mwaldstein** [comment]
We have a similar need - we use keycloak as part of an integrated solution and making 1st instance start up incredibly fast with correct realms/clients in place as fast as the server can get started is a huge benefit, without adding lots of waiting loops or 'waitfor' docker startup commands.

**Comment by DGuhr** [comment]
@pedroigor I also think we should make it available again also for all launch modes (dev/production). In more modern development lifecycles, you really just want to "try out locally" things with devmode, and as far as other envs are concerned, follow the current best practice to have them as close to prod as possible, if not equal. So you'd use the prodmode everywhere, or at least I would. so it's more or less a "black or white" thing it seems. 

tl;dr: For me the question "make it available" is a definite "yes". 

Follow up question is more a "how to achieve the capability (without impacting runtime)" thing. I mentioned the config-cli some times in the past when we talked to @andreaTP and @vmuzikar around operator plans, as it is a great solution we used for some past projects (thanks @jkroepke btw). So it's more a question of "make it a first class citizen, e.g. by adding it to the tools area in the cli" or "use the wildfly way" to me.

**Comment by jkroepke** [comment]
Indeed, we had this need 2017. While import a simple realm was possible, import a single realm and merge it with an existing realm was never possible out of the box. We had pains to describe the flow for operations what they have to click in the UI. Thats the history of [adorsys/keycloak-config-cli](https://github.com/adorsys/keycloak-config-cli).

I also was thinking about an [operator](https://github.com/adorsys/keycloak-config-cli/commit/da5a3a538b87facac68bfb91833448b279a8a340) approach, but I have no clue how much worth is it to develop that, if Keycloak.X is [advertising](https://www.keycloak.org/2021/10/keycloak-x-update) as GitOps friendly too. And my java knowledge lags here a lot.

The transition, to move config from DB to YAML is in my opinion the long time solution.

**Comment by pedroigor** [comment]
Looks like we have a majority then.

FYI, right now we do allow importing realms but you would need to do it in separate steps by running the `import` command. Not necessarily using the API.

But for better UX and automation, perhaps we can have something like a `--import-realm-file` option specific for the start and start-dev commands?

**Comment by pedroigor** [comment]
@MrBuddyCasino We already have the tools to resolve placeholders from env vars. I would say we can consider this requirement.

**Comment by pedroigor** [comment]
Btw, did you try to run `import` command and then `start`? I know not the best but might help while it is not available.

**Comment by eabykov** [comment]
> Btw, did you try to run `import` command and then `start`? I know not the best but might help while it is not available.

@pedroigor yes, I did, it works even if the server is running now, the command just comes out with an error code that it can't use the port and after that, you need to restart the server so that realm appears in the UI

**Comment by eabykov** [comment]
@pedroigor I described this method above in the issue body

**Comment by pedroigor** [comment]
Would make more sense to read any file from an specific directory and then enable import during start-up?

I see this approach less error-prone because you don't need to specify files after mounting them. You just mount in an specific directory like `data/import` and then enable the import by a `--import-realms` option. Making the container spec less verbose and simpler.

The option should be available for both `start` and `start-dev` commands.

**Comment by ruckc** [comment]
I am more likely to manually create a realm for dev... and i would use an export of that for prod.  In addition, IMHO i'd want it to import the realm every startup, incase it changes.  Otherwise configuration drift happens and bam, prod/dev have different state.

**Comment by DGuhr** [comment]
> I am more likely to manually create a realm for dev... and i would use an export of that for prod. In addition, IMHO i'd want it to import the realm every startup, incase it changes. Otherwise configuration drift happens and bam, prod/dev have different state.

Hum, amakes sense for me. So maybe qe have two params here,

```
kc.sh start/start-dev --import-enabled=<once|always> --import-dir=<path:defaulttodata/import>
```
wdyt @pedroigor @stianst

**Comment by eabykov** [comment]
> > I am more likely to manually create a realm for dev... and i would use an export of that for prod. In addition, IMHO i'd want it to import the realm every startup, incase it changes. Otherwise configuration drift happens and bam, prod/dev have different state.
> 
> Hum, makes sense for me. So maybe we have two params here,
> 
> ```
> kc.sh start/start-dev --import=<once|always> --import-dir=<path:defaulttodata/import>
> ```
> 
> wdyt @pedroigor @stianst

@DGuhr can I clarify, will you also be adding environment variables?
- `--import` -> `KC_IMPORT`
- `--import-dir` -> `KC_IMPORT_DIR`

Another question, if you have already done all the variables of the form `KC_` why are variables `KEYCLOAK_ADMIN` and `KEYCLOAK_ADMIN_PASSWORD` still in the old format?

**Comment by DGuhr** [comment]
@eabykov Yes, also an env var for sure. And for KEYCLOAK_ADMIN/ADMIN_PASSWORD: Good question. I think they were not changed because the whole "initial admin creation" is in discussion at the moment, and multiple options are considered, but nothing decided yet, so they were just left as is for now (see https://github.com/keycloak/keycloak/discussions/8549 ).

**Comment by MrBuddyCasino** [comment]
I just discovered the K8S operator supports the definition of realms / users / clients. Is this the preferred way to bootstrap a Keycloak configuration?

**Comment by Schwaller** [comment]
What would the outcome be with the option --import=always? I would be happy if it imports the realm if the realm does not exist yet and skip fancy options. I also don't see the argument about not doing that in production. As I understand Keycloak seems to have no objections to automatically upgrade the DB if needed (not that i object to this behaviour).

**Comment by eabykov** [comment]
> For now i guess we can workaround this change in docker-compose setups by adding an additional container just for importing:
> 
> ```
>   keycloak-initial-config-importer:
>     image: quay.io/keycloak/keycloak:17.0.0
>     depends_on:
>       - "postgres-keycloak"
>     environment:
>       KC_DB_URL_HOST: "postgres-keycloak"
>       KC_DB_URL_PORT: "5432"
>       KC_DB_USERNAME: "keycloak"
>       KC_DB_PASSWORD: "keycloak"
>       KEYCLOAK_ADMIN: admin
>       KEYCLOAK_ADMIN_PASSWORD: admin
> 
>     volumes:
>       - ./keycloak/config:/tmp/keycloak/config
>     command: import --file /tmp/keycloak/config/example-initial-keycloak-config.json
> ```
> 
> How bad is the fact that it will execute multiple times? What will the result be?

Hi @Schwaller, as a result, you will have another application running in import mode

## In order for the container not to restart n times
1. Add to your compose file that options
```yaml
  keycloak-initial-config-importer:
    restart: "no"
    depends_on:
      - keycloak

  keycloak:
    healthcheck:
      test: ["CMD", "curl", "--fail", "http://localhost:8080"]
```
2. Start the [Docker Compose v2](https://docs.docker.com/compose/cli-command/) project with the command `docker compose up -d --wait`

**Comment by Schwaller** [comment]
Hi @eabykov - thx for the quick response! As I understand it the 'restart: "no"' is the default and has not the desired effect (source https://github.com/compose-spec/compose-spec/blob/master/spec.md). Do i miss something? I'm also kinda confused by the healthcheck you suggested. Could you elaborate on that?  

Maybe this is all related to using the V2 of docker-compose spec? I think our team would not like that very much :/ still thanks for quick input!

**Comment by Schwaller** [comment]
In any case if you have the same issue - replacing the entrypoint of the docker image with the following script worked nicely:

```
#!/bin/bash

IMPORT_FLAG_FILE=/opt/keycloak/imported.flag

if [[ -z "${KEYCLOAK_IMPORT}" ]]; then
  echo "Skipping Import (KEYCLOAK_IMPORT not used)"
else
  if test -f "$IMPORT_FLAG_FILE"; then
      echo "Skipping Import (already imported)"
  else
      echo "Import initial config: $KEYCLOAK_IMPORT)"
      /opt/keycloak/bin/kc.sh import --file "$KEYCLOAK_IMPORT"
      touch "$IMPORT_FLAG_FILE"
  fi
fi

echo "Starting Keycloak ... "
/opt/keycloak/bin/kc.sh "$@"

```

When you already at creating your own image its probably worth switching to Alpine as well to avoid numerous security issues with the ubi-8 base.

**Comment by eabykov** [comment]
Hi @Schwaller 
- `I think our team would not like that very much :/` - why your team doesn't like Compose v2?
- `I'm also kinda confused by the healthcheck you suggested` - it is necessary for the `keycloak-initial-config-importer` to start only after the `keycloak` is started

The logic, in this case, is as follows:

Keycloak starts -> keycloak-initial-config-importer wait for healthcheck of Keycloak will be healthy -> keycloak-initial-config-importer starts

**Comment by eabykov** [comment]
Hi @Schwaller 
- `I think our team would not like that very much :/` - why your team doesn't like Compose v2?
- `I'm also kinda confused by the healthcheck you suggested` - it is necessary for the `keycloak-initial-config-importer` to start only after the `keycloak` is started

The logic, in this case, is as follows:

Keycloak starts -> keycloak-initial-config-importer wait for healthcheck of Keycloak will be healthy -> keycloak-initial-config-importer starts

**Comment by eabykov** [comment]
Hi @Schwaller 
- `I think our team would not like that very much :/` - why your team doesn't like Compose v2?
- `I'm also kinda confused by the healthcheck you suggested` - it is necessary for the `keycloak-initial-config-importer` to start only after the `keycloak` is started

The logic, in this case, is as follows:

Keycloak starts -> keycloak-initial-config-importer wait for healthcheck of Keycloak will be healthy -> keycloak-initial-config-importer starts

**Comment by Schwaller** [comment]
Hi @eabykov 

Guess it was a bit late for me yesterday ;)  I thought you were referring to docker compose spec v2.0 not the new docker compose which is part of the docker command now. Anyway, thanks for the clarification :-) What would the effect be of multiple imports? Would it reset settings in some way? 

We will make our own image based on the mentioned entrypoint script above. It has the advantage that we can go Alpine based and therefore avoid a lot of security flags and the fact that we need an Apple M1 compatible version. Would be amazing if Keycloak would go that direction but i'm guessing that would politically a big move ;)

**Comment by eabykov** [comment]
> What would the effect be of multiple imports? Would it reset settings in some way?

@Schwaller it seems that by default realm will be overwritten to prevent this from happening

```shell
docker run --rm quay.io/keycloak/keycloak:17.0.0 import --help
```
```
Import data from a directory or a file.

Usage:

kc.sh import [OPTIONS]

Import data from a directory or a file.

Options:

--dir <path>         Set the path to a directory where files will be created with the exported data.
--file <path>        Set the path to a file that will be created with the exported data.
-h, --help           This help message.
--override <false>   Set if existing data should be skipped or overridden. Default: true.
--realm <realm>      Set the name of the realm to export
```


> It has the advantage that we can go Alpine based

I also wonder why Keycloak is not used as a base image of a small and safe alpine. @DGuhr @pedroigor why?

**Comment by Schwaller** [comment]
Regarding why Alpine is better you can do a "docker scan" - less stuff less security issues basically. Also its tiny and fast :) Question should be why use anything else ;-)

I understand we could always execute a "docker run" but its just an extra step of complication that we have to communicate to our devs, customers etc. Easy is always better.

**Comment by eabykov** [comment]
> Regarding why Alpine is better you can do a "docker scan" - less stuff less security issues basically. Also its tiny and fast :) Question should be why use anything else ;-) Probably boils down to the fact that Keycloak is 'owned' by Redhat and it's a matter of pride not to use a Non-Redhat distro.
> 
> I understand we could always execute a "docker run" but its just an extra step of complication that we have to communicate to our devs, customers etc. Easy is always better. Thank you very much for the --override hint!

@Schwaller try https://github.com/aquasecurity/trivy instead of docker scan

**Comment by eabykov** [comment]
@Schwaller @DGuhr current list of vulnerabilities on the `redhat 8.5` full report [keycloak_report.txt](https://github.com/keycloak/keycloak/files/8148334/keycloak_report.txt)
```
quay.io/keycloak/keycloak:17.0.0 (redhat 8.5)
=============================================
Total: 101 (UNKNOWN: 0, LOW: 37, MEDIUM: 62, HIGH: 2, CRITICAL: 0)
```

For example list of vulnerabilities of `alpine:3.15`: 
```
alpine:3.15 (alpine 3.15.0)
===========================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
```

The base image of `alpine:3.15` is smaller in size and you would thus reduce the size of the Keycloak image
```
REPOSITORY                  TAG           IMAGE ID       CREATED        SIZE
quay.io/keycloak/keycloak   17.0.0        81417e4be7ec   2 weeks ago    556MB
```
> I think there is no need to explain here why it is necessary?

[Trivy](https://github.com/aquasecurity/trivy) was used as a scanning tool

**Comment by madrover** [comment]
What I have seen so far from the [Gitops friendly functionality](https://www.keycloak.org/2021/10/keycloak-x-update) has been focused more on the server configuration aspect but not on the realm configuration aspects which, IMHO, are more complex to handle. 

Ideally we should be handle the KC realms configuration in more declarative way that would enable us to share and promote realms between environments in a more user friendly manner. Currently to make that happen we need to do a bunch of calls to several API endpoints to ensure that the required aspects of the Realm configuration are update properly. This was discussed previously at [KEYCLOAK-12395](https://issues.redhat.com/browse/KEYCLOAK-12395), for example.

It seems to me that the functionality described here seems a step towards the declarative management of the Realms configuration, but I'm not sure if this would cover the complete deal.

**Comment by pedroigor** [comment]
@eabykov And all, could you please check https://github.com/keycloak/keycloak/pull/10754?

**Comment by eabykov** [comment]
> ```shell
> /opt/keycloak
> ```

@pedroigor could not find an environment variable such as `KC_IMPORT_REALM` to enable import in environments such as Docker container etc.

**Comment by pedroigor** [comment]
@eabykov This option is not a configuration option that is why it is not available as an env var. It expects no value. Similar to `auto-build`.

**Comment by eabykov** [comment]
> @eabykov This option is not a configuration option that is why it is not available as an env var. It expects no value. Similar to `auto-build`.

@pedroigor  then as far as I can tell everything is fine, I'm waiting for the release of [17.0.1](https://github.com/keycloak/keycloak/milestone/21) to check in practice [here](https://github.com/eabykov/keycloak-compose)

**Comment by pedroigor** [comment]
Thanks all for the feedback and help.

**Comment by shane-js** [comment]
> In any case if you have the same issue - replacing the entrypoint of the docker image with the following script worked nicely:
> 
> ```
> #!/bin/bash
> 
> IMPORT_FLAG_FILE=/opt/keycloak/imported.flag
> 
> if [[ -z "${KEYCLOAK_IMPORT}" ]]; then
>   echo "Skipping Import (KEYCLOAK_IMPORT not used)"
> else
>   if test -f "$IMPORT_FLAG_FILE"; then
>       echo "Skipping Import (already imported)"
>   else
>       echo "Import initial config: $KEYCLOAK_IMPORT)"
>       /opt/keycloak/bin/kc.sh import --file "$KEYCLOAK_IMPORT"
>       touch "$IMPORT_FLAG_FILE"
>   fi
> fi
> 
> echo "Starting Keycloak ... "
> /opt/keycloak/bin/kc.sh "$@"
> ```
> 
> When you already at creating your own image its probably worth switching to Alpine as well to avoid numerous security issues with the ubi-8 base.

@Schwaller Few questions please: Did this modified entry file end up working for you? If so what is meant by `IMPORT_FLAG_FILE=/opt/keycloak/imported.flag`?



> > @eabykov This option is not a configuration option that is why it is not available as an env var. It expects no value. Similar to `auto-build`.
> 
> @pedroigor then as far as I can tell everything is fine, I'm waiting for the release of [17.0.1](https://github.com/keycloak/keycloak/milestone/21) to check in practice [here](https://github.com/eabykov/keycloak-compose)

Besides the modified entry file - as someone new to Keycloak (running 17.0.1) between this and issue #10754 I am little fuzzy on what's the work around as 10754 states it will be in version 18 yet @eabykov mentioned waiting for 17.0.1 to us it in his example linked.  Is it correct to say that as of version 17.0.1, importing realms from docker start up from both "start" (not start-dev and using something like: "-Dkeycloak.import=/opt/keycloak/data/import/realm.json") and env variable (KEYCLOAK_IMPORT) is removed?

 If so is it further correct to say the only way to do so using the prod start in 17.0.1 is to run docker exec standalone.sh ... after container starts?

Thanks for any clarification you can provide.

**Comment by LoganKnechtMagicLeap** [comment]
Just chiming in to say that the issue of not being able to import realms and users for docker compose has impacted my team.

We had to upgrade to 17.0.1 in order to address Macbook M1 chip concerns, however this feature regression is directly impacting our local development, and testing environments.

If someone has an example on how to stand up the `quay.io/keycloak/keycloak:17.0.1` image with a realm and user in a docker compose configuration, that would greatly be appreciated.

**Comment by Schwaller** [comment]
If the Dockerfile can be improved I would love to hear your suggestions - was a long day ^^

**Comment by LoganKnechtMagicLeap** [comment]
Hahaha @Schwaller - thank you so much for the example!

It's a little different than what I ended up on, so here is my solution to compare and contrast 🙃 

# Create a Valid Keycloak Realm and Export It

Follow the directions listed in the documentation here:
https://www.keycloak.org/getting-started/getting-started-docker

Run this command and create a `Keycloak` instance

```bash
docker run -p 8080:8080 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin quay.io/keycloak/keycloak:17.0.1 start-dev
```

Then go to the url 
http://localhost/admin/

Create a new realm

And in that realm create any of the required pieces you need

- Cients
- Client Scopes
- etc

Once that is done - export it by going selecting the `export` menu on the left

Save that to a directory which will be used in the docker configuration.

**WARNING** This does NOT export users.

# Create User(s)

You will have to manually add users yourself by inspecting the `json` output and adding a specific section like this in the realm configuration

```json
{
  ...
  "realm": "example-realm",
  ...
  "users": [
    {
      "username": "test@example.com",
      "email": "test@example.com",
      "firstName": "Test",
      "lastName": "User",
      "enabled": true,
      "emailVerified": true,
      "credentials": [
        {
          "type": "password",
          "value": "password"
        }
      ],
      "realmRoles": [],
      "clientRoles": {
        "realm-management": []
      }
    }
  ]

}
```

# Configure Directory structure

I have a directory like this

```
.
├── src
│   └── package.json
├── keycloak
│    ├── Dockerfile
│    └── realm-and-users.json
└── ...
```

I place the `realm-and-users.json` in a `keycloak` directory, and then I created a `Dockerfile` in there as well in order to import the realm ahead of time.

# Configure a Docker File to Provision These Pieces Dynamically

To create a dynamic image that will automatically import the `keycloak` realm follow these steps.

```Dockerfile
FROM quay.io/keycloak/keycloak:17.0.1

# Make the realm configuration available for import
COPY realm-and-users.json /opt/keycloak_import/

# Import the realm and user
RUN /opt/keycloak/bin/kc.sh import --file /opt/keycloak_import/realm-and-users.json

# The Keycloak server is configured to listen on port 8080
EXPOSE 8080
EXPOSE 8443

# Import the MagicLeap realm on start-up
CMD ["start-dev"]
```

And that's it!

# Create an Execution Path

In my `package.json` file I have some scripts set up to automatically create the pieces I need

```json
{
  "name": "example-src",
  ...
  "scripts": {
    ...
    "keycloak:build": "cd ../keycloak && docker build --tag keycloak .",
    "keycloak:start": "docker run --publish 8080:8080 --env KEYCLOAK_ADMIN=admin --env KEYCLOAK_ADMIN_PASSWORD=admin keycloak",
  },
  ...
}
```

# Conclusion

With this set-up you should be able to dynamically create a `keycloak` configuration that has a realm(s) available and users available as well!

Then it's just one more step to integrate this with `docker-compose`

**Comment by Schwaller** [comment]
As I understand it RUN is executed at build time so that could become an issue no?

**Comment by LoganKnechtMagicLeap** [comment]
@Schwaller You are correct! This is helpful if someone is home-rolling this, but it could not be a super durable solution

**Comment by omasseau** [comment]
As it seems this as now been fixed in version 18.0.0, could we have a pointer to the new documentation to import realms on startup using a env variable ? ;)
Does it exist yet ?

**Comment by eabykov** [comment]
> As it seems this as now been fixed in version 18.0.0, could we have a pointer to the new documentation to import realms on startup using a env variable ? ;) Does it exist yet ?

Sure: https://www.keycloak.org/server/importExport

Here, you can find an example: https://github.com/eabykov/keycloak-compose

**Comment by pedroigor** [comment]
@pjgg Could you elaborate a bit more, please? Doesn't the `import-realm` option work for you?

**Comment by pjgg** [comment]
No is not working when my realms are stored in a configMap
Keycloak 18.0.0

DeploymentConfig:

```
- apiVersion: "apps.openshift.io/v1"
  kind: "DeploymentConfig"
  metadata:
    labels:
      scenarioId: "OpenShiftUsingExtensionSecurityResourceIT-1655974151772"
    name: "keycloak"
    namespace: "ts-ezzgbjvebz"
  spec:
    replicas: 1
    selector:
      name: "keycloak"
    strategy:
      type: "Recreate"
    template:
      metadata:
        labels:
          name: "keycloak"
          tsLogWatch: "keycloak"
          scenarioId: "OpenShiftUsingExtensionSecurityResourceIT-1655974151772"
        name: "keycloak"
        namespace: "ts-ezzgbjvebz"
      spec:
        containers:
        - args:
          - "start-dev"
          - "--import-realm"
          env:
          - name: "X509_CA_BUNDLE"
            value: "/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt"
          - name: "JGROUPS_ENCRYPT_PROTOCOL"
            value: "ASYM_ENCRYPT"
          - name: "JGROUPS_CLUSTER_PASSWORD"
            value: "test-jgroups-password"
          - name: "JGROUPS_PING_PROTOCOL"
            value: "openshift.DNS_PING"
          - name: "OPENSHIFT_DNS_PING_SERVICE_NAME"
            value: "keycloak-ping"
          - name: "OPENSHIFT_DNS_PING_SERVICE_PORT"
            value: "8888"
          - name: "KEYCLOAK_ADMIN_PASSWORD"
            value: "admin"
          - name: "KEYCLOAK_ADMIN"
            value: "admin"
          image: ""
          imagePullPolicy: "Always"
          name: "keycloak"
          ports:
          - containerPort: 8778
            name: "jolokia"
            protocol: "TCP"
          - containerPort: 8080
            name: "http"
            protocol: "TCP"
          - containerPort: 8443
            name: "https"
            protocol: "TCP"
          - containerPort: 8888
            name: "ping"
            protocol: "TCP"
          resources:
            limits:
              memory: "1Gi"
          volumeMounts:
          - mountPath: "/etc/x509/https"
            name: "keycloak-cert-https-volume"
            readOnly: true
          - mountPath: "/etc/x509/jgroups"
            name: "keycloak-cert-jgroups-volume"
            readOnly: true
          - mountPath: "/opt/keycloak/data/import"
            name: "opt-keycloak-data-import"
            readOnly: true
        terminationGracePeriodSeconds: 75
        volumes:
        - name: "keycloak-cert-https-volume"
          secret:
            secretName: "keycloak-cert-https-secret"
        - name: "keycloak-cert-jgroups-volume"
          secret:
            secretName: "keycloak-cert-jgroups-secret"
        - configMap:
            name: "opt-keycloak-data-import"
          name: "opt-keycloak-data-import"
    triggers:
    - type: "ConfigChange"
    - imageChangeParams:
        automatic: true
        containerNames:
        - "keycloak"
        from:
          kind: "ImageStreamTag"
          name: "keycloak:latest"
      type: "ImageChange"
```

The most relevant of this deployment config is that we have a configMap `opt-keycloak-data-import` that is supported by a volume:

```
 - mountPath: "/opt/keycloak/data/import"
            name: "opt-keycloak-data-import"
            readOnly: true
```

The error that we got when at keycloak start time is the following one: 

```
10:51:08,771 INFO  [keycloak] 2022-06-23 08:51:07,424 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (development) mode
10:51:08,772 INFO  [keycloak] 2022-06-23 08:51:07,424 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to parse realm configuration file: /opt/keycloak/bin/../data/import/..2022_06_23_08_50_05.2663866994
10:51:08,772 INFO  [keycloak] 2022-06-23 08:51:07,424 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Is a directory
```

And I think, that ... this is because internally `keycloak-realm.json` is a symbolic link to another sub-folder:

```
sh-4.4$ ls -la
total 0
drwxrwsrwx. 3 root 1000940000 87 Jun 23 08:57 .
drwxr-xr-x. 3 root root       20 Jun 23 08:57 ..
drwxr-sr-x. 2 root 1000940000 33 Jun 23 08:57 ..2022_06_23_08_57_55.3203107442
lrwxrwxrwx. 1 root 1000940000 32 Jun 23 08:57 ..data -> ..2022_06_23_08_57_55.3203107442
lrwxrwxrwx. 1 root 1000940000 26 Jun 23 08:57 keycloak-realm.json -> ..data/keycloak-realm.json
```

If I check the content of `keycloak.json` on `/opt/keycloak/data/import/` the content is the expected one:

```
sh-4.4$ cat keycloak-realm.json 
{
  "realm": "test-realm",
  "enabled": true,
  "sslRequired": "none",
  "accessTokenLifespan": 5,
  "roles": {
    "realm": [
      {
        "name": "test-user-role"
      },
      {
        "name": "test-admin-role"
      }
    ]
  },
  "users": [
    {
.
.
.
```

When the realms were loaded by an env (Keycloak 16.1.0), we didn't have this problem at all (we were able to load the realms from a configMap) 

Maybe is a separate issue? WDYT?

**Comment by eabykov** [comment]
> ...what about if we recover the ability to load a realm from an environment variable if Keycloak is running in devMode? :)
> 
> Proposal:
> 
> `KEYCLOAK_DEV_REALM_IMPORT:/tmp/realm.json`
> 
> Currently, I am having some troubles in an E2E testing + OCP/k8s, because with the new realm import way that is implemented on Keycloak 18.0, I have to create a `volume/configMap` in my deploymentConfig. This volume is translated into a symbolic link inside the pod
> 
> ```
> sh-4.4$ ls -ls
> total 0
> 0 lrwxrwxrwx. 1 root 1001080000 26 Jun 22 16:55 keycloak-realm.json -> ..data/keycloak-realm.json
> ```
> 
> ... and the hack doesn't work :(
> 
> Do you think any drawbacks to restoring the environment variable on DevMode?
> 
> CC @pedroigor

Hi @pjgg you need to specify `subPath` in the `volumeMounts` section of container

```
      volumeMounts:
      - name: my-realm
        mountPath: /opt/keycloak/data/import/realm.json
        subPath: realm.json
    volumes:
      - name: my-realm
        configMap:
          name: keycloak-realm
```

**Comment by andreaTP** [comment]
@pjgg the error you are reporting sounds like this:
https://github.com/keycloak/keycloak/issues/11700

**Comment by pjgg** [comment]
Yes!, thanks @andreaTP, @eabykov and @pedroigor

I can confirm that is possible to load a reaml from a configMap! 

snnipet:

```
- mountPath: "/opt/keycloak/data/import/keycloak-realm.json"
            name: "opt-keycloak-data-import"
            readOnly: true
            subPathExpr: "keycloak-realm.json"
```

**Comment by andreaTP** [comment]
I'm still curious about the fact that `subPath` is required, for sure, it slightly changes the semantics on how the configMap gets mounted, cc. @pedroigor @DGuhr

**Comment by pjgg** [comment]
To me the most interesting is that Keycloak 16 was able to read the same configMap without a subpathExpr. (Through `KEYCLOAK_REALM_IMPORT` env). Maybe the implementation of the previous versions supports subfolders... and `import-realm` doesn't, I don't know ...

**Comment by flashboss** [comment]
Are there news? I also find it important to be able to use a variable like KEYCLOAK_REALM_IMPORT

**Comment by pedroigor** [comment]
@pjgg You're right, the `import-realm` option only supports regular files (and `.json` suffixed).

**Comment by wuwangzhang1216** [comment]
@LoganKnechtMagicLeap 
I encountered an error when building a customized image based on your docker file, the error that I got is 
`#7 10.22 2022-09-26 21:14:10,137 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (import_export) mode
#7 10.23 2022-09-26 21:14:10,137 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: java.io.FileNotFoundException: /tmp/keycloak_import/realm-sdp.json (Permission denied)
#7 10.23 2022-09-26 21:14:10,137 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: /tmp/keycloak_import/realm-sdp.json (Permission denied)
#7 10.23 2022-09-26 21:14:10,138 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
------
executor failed running [/bin/sh -c /opt/keycloak/bin/kc.sh import --file /tmp/keycloak_import/realm-sdp.json]: exit code: 1`

**Comment by wuwangzhang1216** [comment]
> 

@Schwaller, thanks for providing the quick answer, I solved the permission issue by directly using kubectl cp but also encountered an error when importing the realm config file. My error message looks like the below:
![image](https://user-images.githubusercontent.com/60851473/192386297-02ffcfd4-ed48-415a-a32a-782f7d8e1fa0.png)

**Comment by mcravtov** [comment]
We have solved this by running as JAVA_OPTS_APPEND and realm was imported with all ENV variables.
-Dkeycloak.migration.action=import -Dkeycloak.migration.provider=singleFile -Dkeycloak.migration.file=/opt/keycloak/data/realm.json -Dkeycloak.migration.strategy=IGORE_EXISTING -Dkeycloak.migration.replace-placeholders=true -Dkeycloak.profile.feature.upload_scripts=enabled

**Comment by tania2** [comment]
Thank very much!

**Comment by rahul-roy1-maersk** [comment]
Hi @eabykov @pjgg one question - what if there is multiple realm import jsons to be imported, how do I use then the subPath var? it seems if i just do this below, it is not enough. I have a realm config map which has more than one realm jsons. Please advise.

`          volumeMounts:
          - name: realm-config-volume
            mountPath: /realm-config
            readOnly: true
          args: ["start", "--import-realm"]
          imagePullPolicy: "IfNotPresent"
        volumes:
        - name: realm-config-volume
          configMap:
            name: keycloak-realms-configmap`
